packages/aws-cdk/lib/docker.ts

```diff
@@ -45,28 +32,27 @@ export async function prepareContainerAsset(assemblyDir: string,
 
   try {
     const ecr = await toolkitInfo.prepareEcrRepository(asset);
-    const latest = `${ecr.repositoryUri}:latest`;
 
-    let loggedIn = false;
-
-    // In CI we try to pull latest first
-    if (ci) {
-      await dockerLogin(toolkitInfo);
-      loggedIn = true;
-
-      try {
-        await shell(['docker', 'pull', latest]);
-      } catch (e) {
-        debug('Failed to pull latest image from ECR repository');
+    // if both repo name and image tag are explicitly defined, we assume the
+    // image is immutable and can skip build & push.
+    if (asset.repositoryName && asset.imageTag) {
+      debug(`checking if ${asset.repositoryName}:${asset.imageTag} already exists`);
+      if (await toolkitInfo.checkEcrImage(asset.repositoryName, asset.imageTag)) {
+        debug(`image already exists, skipping`);
+        return [];
       }
     }
 
+    // we use "latest" for image tag for backwards compatibility with pre-1.21.0 apps.
```

Comment:
We don't need to default this to `latest`. Legacy apps don't care what the tag is, and the change won't show up in template diffs. Either make use the requested tag or calculate a hash (or even a uuid) in some way.

Reply:
I guess you are right if we assume that `latest` was only used for the `--ci` feature which is now removed. I'll get rid of it.

Reply:
OK... Thought about this some more. You are right that the tag doesn't really matter now, but I guess `latest` is a good as any, and has the benefit of preserving the current behavior. If you don't feel strongly about this, I rather leave it like this.

Reply:
This still feels dangerous to me, but I guess the specific circumstances in which it fails will never come up through a combination of which values will and won't be set in practical cases.

* All images to same repo => `imageTag` will always be set. We won't push to `latest`. 
* All images to their own repo => we will push to `latest` but then retrieve the `sha256` after pushing and use that to reference the image.

What I'm worried about is "all images to same repo => push to latest" happening, but I guess it won't. Also the 2 paths through the code are not at all clear. Maybe it's easier to just make 2 different methods have obviously different code paths? Trying to combine them makes both paths unclear (to me) and introduces risk of unwanted side effects for future changes.

Reply:
Left the current code path in tact